### ISSUE TITLE
Support Docker compose

### ISSUE THREAD

**Issue by githubsaturn** [caprover/caprover#2175]
**Is your feature request related to a problem? Please describe.**
I know there has been many requests on supporting Docker Compose. But the more I think about it, the more I realize Docker Compose is such a bad model for deploying apps that will be a recipe for disaster. 

Imagine this case:

```yml
version: '3.8'

services:
  nginx:
    image: nginx:latest
    ports:
      - "8080:80"
    volumes:
      - ./conf.d:/etc/nginx/conf.d
```

In this case, you want to deploy the content of `./conf.d` in your local directory to CapRover. You may want to make a change, redeploy it again and you expect the new changes to be reflected.

Now, let's review another one:

```yml
version: '3.8'

services:
  mysql:
    image: my-image-uploader-app
    ports:
      - "8080:80"
    volumes:
      - ./uploaded-photos:/app/uploaded-photos
```

In this case, you DO NOT want the content of `./uploaded-photos` to be uploaded to your server. You just want to redeploy the image (and potentially the env values) when you deploy the new version to CapRover. 

So here is the challenge! Which model should we follow? The second model is obviously the "portable" one that that is what is used by One Click apps at this point. But it completely contradicts with the first model - going back to my original statement that "Docker compose is great to get the skeleton up, but for future deploys it should NOT be used."

Wanted to open this thread to hear the thoughts.

**Comment by thgh** [comment]
I would expect it to match `docker stack deploy` behavior, which is the way to deploy compose files.
https://docs.docker.com/reference/cli/docker/stack/deploy/

Could be challenging as I think lots of things are handled by the docker cli, not just docker engine.

**Comment by githubsaturn** [comment]
> Could be challenging as I think lots of things are handled by the docker cli, not just docker engine.

Yes, this is a separate challenge which is a really big one.


However, even with `docker stack deploy` the behavior from "my machine" to "remote machine" is different - the examples mentioned above highlight the challenge. Basic `./conf.d` and `./uploaded-photos` are completely identical in syntax. But from the user POV, they are supposed to work differently.

**Comment by thgh** [comment]
I would not expect conf.d or uploaded-photos to be uploaded. If uploading folders was a feature, it could be asked interactively and then saved in a configuration file.

Or even better: add service labels to indicate the intention of the developer.

**Comment by githubsaturn** [comment]
> I would not expect conf.d or uploaded-photos to be uploaded

Unfortunately that's not always true. When people ask for Docker Compose support, they expect to be able to deploy a github repo that has a docker-compose file automatically - which is a fair ask. Take [this example](https://github.com/InvoicePlane/InvoicePlane/blob/bf6ddb92bdd522c88376b1e50bec0e48672383cb/docker-compose.yml#L57-L60). `./resources/docker/nginx/invoiceplane.conf` must be copied over or else the app doesn't work. In the same file, there is a place where the [referenced volume](https://github.com/InvoicePlane/InvoicePlane/blob/bf6ddb92bdd522c88376b1e50bec0e48672383cb/docker-compose.yml#L103) is just an empty directory meant NOT to override when re-deploying. 


So basically, that has to be done on volume by volume bases, not even at the service level. My meta point here is that this is a backward step. It all seems to point me to the direction that Docker Compose can be used for the initial deploy, but "re-deploy" compose files just does not make much sense (without a ton of additional configs - volume is just one of them).

**Comment by gdl-reif** [comment]
IMO if I want files to be deployed, I'd include them in my repo. For non-ephemeral files, I'd use volumes which must be defined once and re-used on redeploy (as usual with compose). 

In a cluster or even on a single host, folder mappings are expected to be machine-specific, so no need to deploy anything. At least we work like that (so I might be biased), and we use this approach with tons of compose stacks for our services.

**Comment by khromov** [comment]
+1 to what @thgh said, I have zero expectation of anything being uploaded when using docker-compose with `volumes`, of course both examples would only create a volume mount. It should basically map to the current Volumes config in CapRover. 

If you want to add files you need to write a Dockerfile based on `nginx:latest` and add them yourself.

**Comment by githubsaturn** [comment]
Like I said, if we go down this approach, we cannot use `docker-compose` for repositories like [InvoiceNinja](https://github.com/InvoicePlane/InvoicePlane/blob/bf6ddb92bdd522c88376b1e50bec0e48672383cb/docker-compose.yml#L57-L60)


If we don't include the volumes, then, it only be used for the initial deployment, right? Basically, you just deploy the "docker compose" file and nothing else.

**Comment by gdl-reif** [comment]
@githubsaturn exactly. That's how compose works "standalone" as well, so my expectation would be no different from that.

**Comment by dkyeremeh** [comment]
> Then this is actually fairly simple. The "standalone" version is already [somewhat supported](https://caprover.com/docs/docker-compose.html#how-to-run-docker-compose-on-caprover). All we need to do is to make it more clear and have a clear entrypoint, and some other minor modifications (e.g. not having to add `caproverOneClickApp` , `instruction` etc) .

Exactly. If someone want to see the files in their project, they should include it in their image. Compose in caprover should work exactly like how Docker compose works. No file uploading for volumes
